From 408840844a42b47f4646e2f2b68ecabe3b038987 Mon Sep 17 00:00:00 2001
From: Yu Jiahua <yujiahua1@huawei.com>
Date: Sat, 11 Sep 2021 02:19:44 -0800
Subject: [PATCH 3321/4585] [Huawei] sched: Aware multi-core system for
 optimize loadtracking

Offering: HULK
hulk inclusion
category: feature
bugzilla: 181656

--------------------------------

Optimize load tracking feature have uncertain impact
on scheduler in multi-core system. Therefore, an aware
switch is needed to percept the number of cpus on system,
if one than one cpu is detected, optimize load tracking
feature will be disable.

Signed-off-by: Yu Jiahua <yujiahua1@huawei.com>
Reviewed-by: Chen Hui <judy.chenhui@huawei.com>
Signed-off-by: Chen Jun <chenjun102@huawei.com>
Patchwork Links: http://patchwork.huawei.com/project/hulk5.10/list/?series=18808
Signed-off-by: Zheng Zucheng <zhengzucheng@huawei.com>
---
 include/linux/sched/sysctl.h |  6 +--
 kernel/sched/fair.c          | 74 +++++++++++++++++++-----------------
 kernel/sysctl.c              |  6 ++-
 3 files changed, 46 insertions(+), 40 deletions(-)

diff --git a/include/linux/sched/sysctl.h b/include/linux/sched/sysctl.h
index eb2a2d8ace8d..380f6cb6bbfe 100644
--- a/include/linux/sched/sysctl.h
+++ b/include/linux/sched/sysctl.h
@@ -112,12 +112,12 @@ extern int sysctl_blocked_averages(struct ctl_table *table, int write,
 		  void __user *buffer, size_t *lenp, loff_t *ppos);
 extern int sysctl_tick_update_load(struct ctl_table *table, int write,
 		  void __user *buffer, size_t *lenp, loff_t *ppos);
-extern int sysctl_update_load_latency(struct ctl_table *table, int write,
-		  void __user *buffer, size_t *lenp, loff_t *ppos);
+extern int sysctl_update_load_tracking_aware(struct ctl_table *table,
+		  int write, void __user *buffer, size_t *lenp, loff_t *ppos);
 
-extern unsigned int sysctl_load_tracking_latency;
 extern struct static_key_true sched_tick_update_load;
 extern struct static_key_true sched_blocked_averages;
+extern struct static_key_false sched_load_tracking_aware_enable;
 #endif
 
 #endif /* _LINUX_SCHED_SYSCTL_H */
diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index ab8b7dd99385..40b30a7e40fc 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -46,34 +46,35 @@ unsigned int sysctl_sched_latency			= 6000000ULL;
 static unsigned int normalized_sysctl_sched_latency	= 6000000ULL;
 
 #ifdef CONFIG_IAS_SMART_LOAD_TRACKING
-#define LANTENCY_MIN 10
-#define LANTENCY_MAX 30
-unsigned int sysctl_load_tracking_latency = LANTENCY_MIN;
+DEFINE_STATIC_KEY_FALSE(sched_load_tracking_aware_enable);
+static void set_load_tracking_aware(bool enabled)
+{
+	if (enabled)
+		static_branch_enable(&sched_load_tracking_aware_enable);
+	else
+		static_branch_disable(&sched_load_tracking_aware_enable);
+}
 
-int sysctl_update_load_latency(struct ctl_table *table, int write,
-			       void __user *buffer, size_t *lenp, loff_t *ppos)
+int sysctl_update_load_tracking_aware(struct ctl_table *table, int write,
+			   void __user *buffer, size_t *lenp, loff_t *ppos)
 {
-	int ret;
-	int min = LANTENCY_MIN;
-	int max = LANTENCY_MAX;
-	int latency = sysctl_load_tracking_latency;
 	struct ctl_table t;
+	int err;
+	int state = static_branch_likely(&sched_load_tracking_aware_enable);
 
 	if (write && !capable(CAP_SYS_ADMIN))
 		return -EPERM;
 
 	t = *table;
-	t.data = &latency;
-	t.extra1 = &min;
-	t.extra2 = &max;
-
-	ret = proc_dointvec_minmax(&t, write, buffer, lenp, ppos);
-	if (ret || !write)
-		return ret;
+	t.data = &state;
+	err = proc_dointvec_minmax(&t, write, buffer, lenp, ppos);
+	if (err < 0)
+		return err;
 
-	sysctl_load_tracking_latency = latency;
+	if (write)
+		set_load_tracking_aware(state);
 
-	return 0;
+	return err;
 }
 #endif
 
@@ -3858,39 +3859,42 @@ static void detach_entity_load_avg(struct cfs_rq *cfs_rq, struct sched_entity *s
 #define SKIP_AGE_LOAD	0x2
 #define DO_ATTACH	0x4
 
+#ifdef CONFIG_IAS_SMART_LOAD_TRACKING
+/*
+ * Check load tracking senario. In single-core system without cpu frequency update,
+ * precise load tracking will be unnecessary. So here we just shutdown load tracking,
+ * for decreasing cpu usage.
+ */
+static inline int check_load_switch(void)
+{
+	if (static_branch_unlikely(&sched_load_tracking_aware_enable))
+		if (num_online_cpus() == 1)
+		/* no need to update load average in single core senario */
+			return 1;
+
+	return 0;
+}
+#endif
+
 /* Update task and its cfs_rq load average */
 static inline void update_load_avg(struct cfs_rq *cfs_rq, struct sched_entity *se, int flags)
 {
 	u64 now = cfs_rq_clock_pelt(cfs_rq);
 	int decayed;
+
 #ifdef CONFIG_IAS_SMART_LOAD_TRACKING
-	u64 delta;
+	if (check_load_switch())
+		return;
 #endif
-
 	/*
 	 * Track task load average for carrying it to new CPU after migrated, and
 	 * track group sched_entity load average for task_h_load calc in migration
 	 */
-#ifdef CONFIG_IAS_SMART_LOAD_TRACKING
-	delta = now - se->avg.last_update_time;
-	delta >>= sysctl_load_tracking_latency;
-
-	if (!delta)
-		return;
-
-	if (se->avg.last_update_time && !(flags & SKIP_AGE_LOAD))
-		__update_load_avg_se(now, cfs_rq, se);
-
-	decayed  = update_cfs_rq_load_avg(now, cfs_rq);
-	decayed |= propagate_entity_load_avg(se);
-#else
 	if (se->avg.last_update_time && !(flags & SKIP_AGE_LOAD))
 		__update_load_avg_se(now, cfs_rq, se);
 
 	decayed  = update_cfs_rq_load_avg(now, cfs_rq);
 	decayed |= propagate_entity_load_avg(se);
-#endif
-
 
 	if (!se->avg.last_update_time && (flags & DO_ATTACH)) {
 
diff --git a/kernel/sysctl.c b/kernel/sysctl.c
index 84c29ae6f9e7..5b0da262b98f 100644
--- a/kernel/sysctl.c
+++ b/kernel/sysctl.c
@@ -1719,11 +1719,13 @@ static struct ctl_table ias_table[] = {
 		.extra2		= SYSCTL_ONE,
 	},
 	{
-		.procname	= "sched_load_tracking_latency",
+		.procname	= "sched_load_tracking_aware_enable",
 		.data		= NULL,
 		.maxlen		= sizeof(unsigned int),
 		.mode		= 0644,
-		.proc_handler	= sysctl_update_load_latency,
+		.proc_handler	= sysctl_update_load_tracking_aware,
+		.extra1		= SYSCTL_ZERO,
+		.extra2		= SYSCTL_ONE,
 	},
 #endif
 	{ }
-- 
2.17.1

